lib/OpenQA/Utils.pm

```diff
@@ -278,65 +278,85 @@ sub log_fatal {
+# add_log_channel to learn on how to set a channel as default.
 sub _log_msg {
     my ($level, $msg, %options) = @_;
-    my $log_to_standard = 0;
 
+    # use default options
     if (!%options) {
-        # set defaults
-        _log_msg($level, $msg, channels => $log_defaults{CHANNELS}, standard => $log_defaults{LOG_TO_STANDARD_CHANNEL});
-        $log_to_standard = 0;
+        return _log_msg(
+            $level, $msg,
+            channels => $log_defaults{CHANNELS},
+            standard => $log_defaults{LOG_TO_STANDARD_CHANNEL});
     }
-    else {
-        # TODO: We need to get rid of $app here :(
-        $msg = "[pid:$$] " . $msg
-          if defined $app && $app->can('log') && $app->log->can('level') && $app->log->level eq 'debug';
-        if ($options{channels}) {
-            if (ref($options{channels}) eq 'ARRAY') {
-                for my $channel (@{$options{channels}}) {
-                    _log($level, $msg, $channel);
-                }
-            }
-            else {
 
-                _log($level, $msg, $options{channels});
-            }
+    # prepend process ID on debug level
+    if (_current_log_level eq 'debug') {
+        $msg = "[pid:$$] $msg";
+    }
+
+    # log to channels
+    my $wrote_to_at_least_one_channel = 0;
+    if (my $channels = $options{channels}) {
+        for my $channel (ref($channels) eq 'ARRAY' ? @$channels : $channels) {
+            $wrote_to_at_least_one_channel |= _log_to_channel_by_name($level, $msg, $channel);
         }
-        $log_to_standard = $options{standard} // $log_defaults{LOG_TO_STANDARD_CHANNEL};
     }
 
-    _log($level, $msg) if $log_to_standard;
+    # log to standard (as fallback or when explicitely requested)
+    if (!$wrote_to_at_least_one_channel || ($options{standard} // $log_defaults{LOG_TO_STANDARD_CHANNEL})) {
+        # use Mojolicious app if available and otherwise just STDERR/STDOUT
+        _log_via_mojo_app($level, $msg) or _log_to_stderr_or_stdout($level, $msg);
+    }
+}
+
+sub _log_to_channel_by_name {
+    my ($level, $msg, $channel_name) = @_;
+
+    return 0 unless ($channel_name);
+    my $channel = $channels{$channel_name} or return 0;
+    return _try_logging_to_channel($level, $msg, $channel);
+}
+
+sub _log_via_mojo_app {
+    my ($level, $msg) = @_;
+
+    return 0 unless ($app && $app->log);
+    return _try_logging_to_channel($level, $msg, $app->log);
 }
 
-# There are three possibilities for logging:
-# 1- Logging to a channel
-# 2- Logging to the default destination
-# 3- Logging to the STDERR/STDOUT as a fallback in case of none of the above are set
-sub _log {
+sub _try_logging_to_channel {
     my ($level, $msg, $channel) = @_;
 
-    if ($channel && $channels{$channel}) {
-        $channels{$channel}->$level($msg);
+    my $success = 0;
+    try {
+        $channel->$level($msg);
+        $success = 1;
     }
-    elsif ($app && $app->log) {
-        $app->log->$level($msg);
+    catch {
```

Comment:
Why the empty catch statement? would not be better to use it to detect the failure ? (or alternatively, just eval ?)

Reply:
changed to `eval`